KIPs/kip-149.md

```diff
@@ -0,0 +1,292 @@
+
+#### Chain Configuration
+
+In the Chain Config, the following field is introduced. All node operators in a network must update `genesis.json` configuration with the same value. The configuration values for Baobab and Cypress networks are hard-coded on the client source code.
+
+- `RegistryAddress`: the reserved address for the Registry contract, which is `0x0000000000000000000000000000000000000401`.
+
+- `Kip149CompatibleBlock`: the target block number that the Registry will be deployed.
+
+- `RegistryConfig`: the initial data config for the Registry. It is used when injecting the initial state after deploying the Registry.
+  
+```go
+// In klaytn/params/config.go
+type RegistryConfig struct {
+	Records  map[string]common.Address // Map for system contracts
+	Owner   common.Address // Address for initial owner of Registry
+}
+
+var ChainConfig = &ChainConfig{
+    ...
+    Kip149CompatibleBlock: TBD,
+    // RegistryInit will be used when injecting initial state for the Registry
+    // Note that the registry only records the system contracts based on the KIP-149, which is currently only KIP-113
+    // The activation block deployed by state injection will be 0
+    RegistryInit: &RegistryConfig{
+		Records: map[string]common.Address{
+			"KIP113": Kip113Address,
+		},
+		Owner: OwnerAddress,
+    },
+}
+```
+
+
+#### Execution
+
+The Registry deployment is executed at the `engine.Finalize` function, which means the end of the block generation. It reads the reserved address and runtime bytecode and deploys the Registry by the bytecode injection. Also, it injects the initial state provided by `RegistryConfig` for the Registry here. When `kip149CompatibleBlock == 0`, the Registry will be allocated at the genesis block.
+
+```go
+if chain.Config().IsKIP149ForkBlock(header.Number) {
+	// Inject the bytecode and states for the Registry
+	err := registry.InstallRegistry(state, chain.Config().RegistryInit)
+	if err != nil {
+		logger.Error("failed to set the registry contract code", "err", err)
+	} else {
+		logger.Info("successfully set the registry contract code", "block", header.Number.Uint64())
+	}
+}
+```
+
+#### Resolver
+
+A Klaytn node will have a resolver to read the active addresses of system contracts from the Registry. But the system contracts deployed by state injection (e.g., `KIP113`) will be directly read from `chain.Config().RegistryInit` at the `Kip149CompatibleBlock` since the Registry will be deployed in the `engine.Finalize` function, which is the last part of the block generation. In other words, the resolver will be used starting in the `Kip149CompatibleBlock + 1`.
+
+```go
+// Note that it will be used starting in the next of kip-149 fork block
+func ReadRegistryActiveAddr(backend bind.ContractCaller, name string, num *big.Int) (common.Address, error) {
+	code, err := backend.CodeAt(context.Background(), RegistryAddr, num)
```

Comment:
I cannot see any code that this function will be used after Kip149CompatibleBlock.

And the function name is not clear. Maybe `ReadActiveAddressFromRegistry`?

Reply:
Also PR for Klaytn core: https://github.com/klaytn/klaytn/pull/2061